### PR TITLE
Add OnlyAvailable and PopularFirst options for featured products

### PR DESCRIPTION
**Pull request by kelostrada**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Adds an ability to change order of the featured products and filter out products that are not available for ordering. Right now if a product is unavailable it can still be visible inside featured products which can be kinda pointless. This PR will also introduce a way to change the order of displayed products to display products which have been ordered the most.
| Type?         | improvement
| BC breaks?    |  no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | Change module settings, Verify homepage

image

### CODE REVIEW

**Review by matthieu-rolland — COMMENTED** [review]
@PrestaShop/product-team for your information, this is a new feature

**Review by Hlavtox — COMMENTED** [review]
I think you should implement a featured products search provider instead of hardcoding the query. WDYT @kpodemski

**Comment by kelostrada** [comment]
Good point, I mostly left it here for any1 who would want to have a feature like this. Might rework this according to the tip, thanks @Hlavtox

File: ps_featuredproducts.php [hunk 1]
```diff
@@ -263,37 +314,94 @@ public function getWidgetVariables($hookName = null, array $configuration = [])
...
 
-        if (Configuration::get('HOME_FEATURED_RANDOMIZE')) {
-            $query->setSortOrder(SortOrder::random());
+        $random = !!Configuration::get('HOME_FEATURED_RANDOMIZE');
+        $popularFirst = !!Configuration::get('HOME_FEATURED_POPULAR_FIRST');
+        $onlyAvailable = !!Configuration::get('HOME_FEATURED_ONLY_AVAILABLE');
+
+        $outOfStockGlobal = (int) Configuration::get("PS_ORDER_OUT_OF_STOCK");
+
+        $sql = 'SELECT p.*, o.orders_placed, product_shop.*, stock.out_of_stock, IFNULL(stock.quantity, 0) AS quantity' . (Combination::isFeatureActive() ? ', IFNULL(product_attribute_shop.id_product_attribute, 0) AS id_product_attribute,
```
Comment (kpodemski) [inline ps_featuredproducts.php:335]: @kelostrada, there's really no way to query everything like this, Presenters handle getting images, etc.

File: ps_featuredproducts.php [hunk 2]
```diff
@@ -263,37 +314,94 @@ public function getWidgetVariables($hookName = null, array $configuration = [])
...
+        if (!$result) {
+            $result = [];
         }
 
-        $result = $searchProvider->runQuery(
-            $context,
-            $query
-        );
+        // Modify SQL result
+        $products = Product::getProductsProperties($this->context->language->id, $result);
```
Comment (kpodemski) [inline ps_featuredproducts.php:404]: @kelostrada again, it is inside Presenter, the only thing you need for Presenter to work are product IDs

File: ps_featuredproducts.php (hunk 1 above)
Comment (kelostrada) [reply ps_featuredproducts.php:335]: Yeah I get it, the query was basically copy pasted from `Category` that is used here by `CategoryProductSearchProvider`. I can clean it up if you want to use this PR. As I said I just wanted to make it work and felt I can share if anyone needs similar functionality

Comment (kpodemski) [reply ps_featuredproducts.php:335]: thanks @kelostrada 

if you are motivated to finish your PR I'll be happy to accept your contribution, but it has to be reworked to keep the codebase clean 👍🏻

**Comment by kpodemski** [comment]
Since we have had no news from you for more than 20 days, I'm closing this PR. Feel free to re-open it if you find the motivation to finish it. Don't hesitate to ping me if you have any questions or doubts.

**Pull request closed without merging**